Make the latest version of WebKit working on QNX

fxing-qnx opened the pull request:
#### f0d351dc818c8328bee171398eab709d0c81cb29
<pre>
Make the latest version of WebKit working on QNX
<a href="https://bugs.webkit.org/show_bug.cgi?id=303626">https://bugs.webkit.org/show_bug.cgi?id=303626</a>

Reviewed by Justin Michaud.

- MemoryPressureHandlerUnix.cpp is now included in CMake for QNX
- Stack bounds for each JIT instance was not correctly set, __tls() only return thread infomation for the calling thread rather than a newly spawned thread
- currentExecutablePath() is now implemented for QNX
- stackPointer() is now implemented for QNX
- structureHeapAddressSize is now set to 128M for QNX. QNX does not allowed lazy allocation of memory, mapping 4G will ends up with allocating 4G
- Fix some headers includes
- numberOfProcessorCores() is now implemented for QNX

(computeJSCBytecodeCacheVersion()):
* Source/JavaScriptCore/runtime/JSCBytecodeCacheVersion.cpp:
(static inline void*&amp; stackPointerImpl()):
* Source/JavaScriptCore/runtime/MachineContext.h:
(structureHeapAddressSize):
* Source/JavaScriptCore/runtime/StructureID.h:
(WTF::numberOfProcessorCores()):
* Source/WTF/wtf/NumberOfCores.cpp:
(WTF::StackBounds::currentThreadStackBoundsInternal()):
* Source/WTF/wtf/PlatformHave.h:
* Source/WTF/wtf/StackBounds.cpp:
(WTF::currentExecutablePath()):
* Source/WTF/wtf/glib/FileSystemGlib.cpp:
(WTF::OSAllocator::commit(...)):
* Source/WTF/wtf/posix/OSAllocatorPOSIX.cpp:
* Source/WTF/wtf/PlatformGTK.cmake:

Canonical link: <a href="https://commits.webkit.org/304481@main">https://commits.webkit.org/304481@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f8b6ae3fd958c3cae80b4dc85eefcabdd9621219

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/135642 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/8016 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/46934 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/143360 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/87308 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/8660 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/7862 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/103658 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/71153 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/138588 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/6247 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/121596 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/84530 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/6015 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/3623 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/3966 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/127635 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/115240 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/39782 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/146106 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/134126 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/7701 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/40351 "Passed tests") | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/7737 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/6471 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28526 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/5873 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/117895 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/61668 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/7751 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36005 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/166958 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/7498 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/71303 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/43586 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/7720 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/7598 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

justinmichaud commented on Source/WTF/wtf/StackBounds.cpp:
```diff
@@ -81,29 +81,29 @@ WTF_ALLOW_UNSAFE_BUFFER_USAGE_END
 
 StackBounds StackBounds::newThreadStackBounds(PlatformThreadHandle thread)
 {
-    struct _thread_local_storage* tls = __tls();
```
I don't quite understand why this doesn't work. This is called from the new jit thread, correct?

fxing-qnx replied:
Yes, but the old one does not. The old one calls `__tls()` inside the creator's thread hence it will return creater's stack bound rather than the newly created thread's stackbound. I've tested that already.

justinmichaud replied:
Ah, I got confused by the diff. LGTM

mcatanzaro commented on Source/WTF/wtf/glib/FileSystemGlib.cpp:
```diff
@@ -76,6 +76,21 @@ CString currentExecutablePath()
+        return { };
+WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN // QNX port
+    return CString(std::span { readBuffer, static_cast<size_t>(result) });
+WTF_ALLOW_UNSAFE_BUFFER_USAGE_END
```
A better way to write the same thing, without `WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN/END`:

```suggestion
    return unsafeMakeSpan(readBuffer, static_cast<size_t>(result));
```

mcatanzaro replied:
Um, it looks like I'm not smart enough to make the GitHub suggestion work properly. That's intended to replace all three lines, not just one.

fxing-qnx replied:
GitHub is a bit weird sometimes, for example that funny diff for stackbounds

mcatanzaro commented on Source/WTF/wtf/PlatformGTK.cmake:
```suggestion
elseif (CMAKE_SYSTEM_NAME MATCHES "FreeBSD" OR CMAKE_SYSTEM_NAME MATCHES "QNX")
```

justinmichaud reviewed (approved):
r=me with @mcatanzaro's suggestions. It would be really cool to see a demo of this running!

fxing-qnx committed:
Make the latest version of WebKit working on QNX
https://bugs.webkit.org/show_bug.cgi?id=303626

Reviewed by Justin Michaud.

- MemoryPressureHandlerUnix.cpp is now included in CMake for QNX
- Stack bounds for each JIT instance was not correctly set, __tls() only return thread infomation for the calling thread rather than a newly spawned thread
- currentExecutablePath() is now implemented for QNX
- stackPointer() is now implemented for QNX
- structureHeapAddressSize is now set to 128M for QNX. QNX does not allowed lazy allocation of memory, mapping 4G will ends up with allocating 4G
- Fix some headers includes
- numberOfProcessorCores() is now implemented for QNX

(computeJSCBytecodeCacheVersion()):
* Source/JavaScriptCore/runtime/JSCBytecodeCacheVersion.cpp:
(static inline void*& stackPointerImpl()):
* Source/JavaScriptCore/runtime/MachineContext.h:
(structureHeapAddressSize):
* Source/JavaScriptCore/runtime/StructureID.h:
(WTF::numberOfProcessorCores()):
* Source/WTF/wtf/NumberOfCores.cpp:
(WTF::StackBounds::currentThreadStackBoundsInternal()):
* Source/WTF/wtf/PlatformHave.h:
* Source/WTF/wtf/StackBounds.cpp:
(WTF::currentExecutablePath()):
* Source/WTF/wtf/glib/FileSystemGlib.cpp:
(WTF::OSAllocator::commit(...)):
* Source/WTF/wtf/posix/OSAllocatorPOSIX.cpp:
* Source/WTF/wtf/PlatformGTK.cmake:

Canonical link: https://commits.webkit.org/304481@main
